### PR TITLE
Task04

### LINKED ISSUE #4: Image Tags
#### As a user I want to add tags to images.

__Story__:

Our users are loving the ability to add images to our service. However, now
that there are more than a handful of images, it is apparent that we need some
way to filter images by certain categories. In order to do that, we first need
to label images in some way. It is time to add the ability for our users to tag
images when they add them.

__Acceptance criteria__:
- [x] When I add a new image, there is a form field to add tags.
- [x] After I've saved an image with tags, I see those tags displayed with the
  associated image.
- [x] My application uses the acts-as-taggable-on gem to provide the tagging
  functionality.

__Reference material__:
- https://github.com/mbleigh/acts-as-taggable-on

__Note__: You should allow images to have no tags; this sets the stage for a later issue.

__Dependencies__:
- Image Index

### PR DESCRIPTION
**Pull request by lei-appfolio**
Closes #4 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.
- [x] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: test/controllers/images_controller_test.rb [hunk 1]
```diff
@@ -8,7 +8,13 @@ def test_new
...
     assert_select '.new_image'
     assert_select '.simple_form'
-    assert_select '#new_image'
+
+    assert_select '#new_image', 1
+    assert_select '#new_image' do
+      assert_select '.image_web_url'
+      assert_select '.image_mytag_list'
+      assert_select '.btn'
+    end
```
Comment (bhargav265) [inline test/controllers/images_controller_test.rb]: Select the class added to the defaults

File: app/controllers/images_controller.rb [hunk 2]
```diff
@@ -27,4 +28,8 @@ def create
   def image_params
     params.require(:image).permit(:web_url)
```
Comment (bhargav265) [inline app/controllers/images_controller.rb]: Permit both the params here

File: test/controllers/images_controller_test.rb [hunk 3]
```diff
@@ -80,6 +80,18 @@ def test_create__succeed
     assert_equal 'Image url was successfully submitted.', flash[:notice]
```
Comment (bhargav265) [inline test/controllers/images_controller_test.rb]: `assert_equal nil, Image.last.mytag_list`

File: test/controllers/images_controller_test.rb [hunk 4]
```diff
@@ -42,19 +74,10 @@ def test_index_image_order
...
-
-    assert_select 'li:nth-child(2)' do
-      assert_select format('img[src="%<url>s"]', url: url2)
-    end
-
-    assert_select 'li:nth-child(3)' do
-      assert_select format('img[src="%<url>s"]', url: url1)
+    assert_select 'img' do |element|
+      assert_equal url3, element[0][:src]
+      assert_equal url2, element[1][:src]
```
Comment (bhargav265) [inline test/controllers/images_controller_test.rb]: 👍

File: test/controllers/images_controller_test.rb (hunk 1 above)
Comment (lei-appfolio) [reply test/controllers/images_controller_test.rb]: OK!

File: app/controllers/images_controller.rb (hunk 2 above)
Comment (lei-appfolio) [reply app/controllers/images_controller.rb]: OK!

File: test/models/image_test.rb [hunk 5]
```diff
@@ -20,4 +20,35 @@ def test_image_url__invalid_if_it_is_wrong_format
...
+    assert_predicate image, :valid?
+    assert_equal ['cute'], image.mytag_list
+  end
+
+  def test_image_tag__multiple
+    image = Image.new(web_url: 'https://i.pinimg.com/474x/b4/bb/eb/b4bbeb2aaafd59040e56df10ad885b40.jpg')
+    image.mytag_list.add('cute', 'panda')
+
+    assert_predicate image, :valid?
+    assert_equal %w[cute panda], image.mytag_list
```
Comment (bhargav265) [inline test/models/image_test.rb]: `assert_sets_equal` can be used in the future. But this is okay

**Pull request merged**